hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/AbstractClientInternalCacheProxy.java

```diff
@@ -157,13 +159,17 @@ protected void onInitialize() {
     public void setCacheManager(HazelcastCacheManager cacheManager) {
         assert cacheManager instanceof HazelcastClientCacheManager;
 
-        this.cacheManager = (HazelcastClientCacheManager) cacheManager;
+        if (!cacheManagerRef.compareAndSet(null, (HazelcastClientCacheManager) cacheManager)) {
+            if (cacheManagerRef.get() != cacheManager) {
```

Comment:
why is this extra check needed? Either you are going to win the cas, and you are happy, or you going to loose the cas and throw an error.